design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,223 @@
+
+## User Stories
+* As a cluster admin, I want to let `isolatedEmulatorThread` VMs with an even number of CPUs to run on my cluster by setting a cluster-wide knob on HCO.
+* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:
+```
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  annotations:
+    kubevirt.io/EmulatorThreadCompleteToEvenParity:
+spec:
+  configuration:
+    developerConfiguration:
+      featureGates:
+      - EmulatorThreadCompleteToEvenParity
+```
+This annotation will be supported only when the `EmulatorThreadCompleteToEvenParity` featureGate is enabled.
+
+## Virt-API:
+
+### VMI mutator webhook
+When the following conditions are met:
+1. The feature gate is enabled on the Kubevirt CR
+2. Kubevirt CR has the annotation.
+3. VMI is configured with `isolateEmulatorThread: true`
+
+then the VMI mutator webhook will
+1. Copy the annotation from the kubevirt CR to the VMI annotations. This is done in order for it to be used on the virt-launcher, when setting the housekeeping cgroup. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.
+
+### VMI create-admitter validation webhook
+Should allow the VMI to have the new annotation only if:
+1. The feature gate is enabled on the Kubevirt CR
+2. VMI is configured with `isolateEmulatorThread: true`
+
+## Virt-controller:
+When the following conditions are met:
+1. VMI is configured with `isolateEmulatorThread: true`
+2. VMI has the annotation.
+
+then virt-controller will:
+1. Set the virt-launcher pod with an even amount of CPUs, meaning:
+
+| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) | Notes                                                                       |
+|-----------------------------------|------------------|-----------------------|-----------------------------------------------------------------------------|
+| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         | * In both scenarios the guest will get the same amount of CPUs it requested |
+
+## Virt-launcher:
+When the following conditions are met:
+1. VMI is configured with `isolateEmulatorThread: true`
+2. VMI has the annotation.
+
+then virt-launcher pod will then:
+1. Assign the emulator Threads (one or two, depending on the User requested CPUs as shown in the table above) to the domain's emulatorThreadPin
```

Comment:
No need for numbering because there is only one step.

Reply:
DONE